sys/net/netstats/netstats_neighbor.c

```diff
@@ -0,0 +1,410 @@
+    }
+
+    stats->last_updated = now;
+}
+
+static bool isfresh(netstats_nb_t *stats)
+{
+    uint16_t now = xtimer_now_usec() / US_PER_SEC;;
+
+    /* Half freshness if applicable to update to current freshness */
+    half_freshness(stats, now);
+
+    return (stats->freshness >= NETSTATS_NB_FRESHNESS_TARGET) &&
+           (now - stats->last_updated < NETSTATS_NB_FRESHNESS_EXPIRATION);
+}
+
+bool netstats_nb_isfresh(netif_t *dev, netstats_nb_t *stats)
+{
+    bool ret;
+
+    _lock(dev);
+    ret = isfresh(stats);
+    _unlock(dev);
+
+    return ret;
+}
+
+void netstats_nb_init(netif_t *dev)
+{
+    mutex_init(&dev->neighbors.lock);
+
+    _lock(dev);
+    memset(dev->neighbors.pstats, 0, sizeof(netstats_nb_t) * NETSTATS_NB_SIZE);
+    cib_init(&dev->neighbors.stats_idx, NETSTATS_NB_QUEUE_SIZE);
+    _unlock(dev);
+}
+
+static void netstats_nb_create(netstats_nb_t *entry, const uint8_t *l2_addr, uint8_t l2_len)
+{
+    memset(entry, 0, sizeof(netstats_nb_t));
+    memcpy(entry->l2_addr, l2_addr, l2_len);
+    entry->l2_addr_len = l2_len;
+
+#ifdef MODULE_NETSTATS_NEIGHBOR_ETX
+    entry->etx = NETSTATS_NB_ETX_INIT * NETSTATS_NB_ETX_DIVISOR;
+#endif
+}
+
+bool netstats_nb_get(netif_t *dev, const uint8_t *l2_addr, uint8_t len, netstats_nb_t *out)
+{
+    _lock(dev);
+
+    netstats_nb_t *stats = dev->neighbors.pstats;
+    bool found = false;
+
+    for (int i = 0; i < NETSTATS_NB_SIZE; i++) {
+
+        /* Check if this is the matching entry */
+        if (l2util_addr_equal(stats[i].l2_addr, stats[i].l2_addr_len, l2_addr, len)) {
+            *out = stats[i];
+            found = true;
+            break;
+        }
+    }
+
+    _unlock(dev);
+    return found;
+}
+
+/* find the oldest inactive entry to replace. Empty entries are infinity old */
+static netstats_nb_t *netstats_nb_get_or_create(netif_t *dev, const uint8_t *l2_addr, uint8_t len)
+{
+    netstats_nb_t *old_entry = NULL;
+    netstats_nb_t *stats = dev->neighbors.pstats;
+    uint16_t now = xtimer_now_usec() / US_PER_SEC;
+
+    for (int i = 0; i < NETSTATS_NB_SIZE; i++) {
+
+        /* Check if this is the matching entry */
+        if (l2util_addr_equal(stats[i].l2_addr, stats[i].l2_addr_len, l2_addr, len)) {
+            return &stats[i];
+        }
+
+        /* Entry is oldest if it is empty */
+        if (stats[i].l2_addr_len == 0) {
+            old_entry = &stats[i];
+            continue;
+        }
+
+        /* Check if the entry is expired */
+        if ((isfresh(&stats[i]))) {
+            continue;
+        }
+
+        /* Entry is oldest if it is expired */
+        if (old_entry == NULL) {
+            old_entry = &stats[i];
+            continue;
+        }
+
+        /* don't replace old entry if there are still empty ones */
+        if (old_entry->l2_addr_len == 0) {
+            continue;
+        }
+
+        /* Check if current entry is older than current oldest entry */
+        old_entry = netstats_nb_comp(old_entry, &stats[i], now);
```

Comment:
You could get rid of all those `continue`s and make the code a bit easier to follow:

```suggestion
        if (stats[i].l2_addr_len == 0) {
            old_entry = &stats[i];
        }
        /* Check if the entry is expired */
        else if (!isfresh(&stats[i])) {
            /* Entry is oldest if it is expired */
            if (old_entry == NULL) {
                old_entry = &stats[i];
            }
            /* don't replace old entry if there are still empty ones */
            else if (old_entry->l2_addr_len > 0) {
                /* Check if current entry is older than current oldest entry */
                old_entry = netstats_nb_comp(old_entry, &stats[i], now);
            }
        }
```

Reply:
I guess that's a matter of personal taste, I find code with *less* nesting easier to follow.